core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -1308,72 +1322,116 @@ public void run() {
+              if (!serviceConfigUpdated) {
+                // First DNS lookup has invalid service config, and cannot fall back to default
                 channelLogger.log(
                     ChannelLogLevel.INFO,
-                    "Received no service config, using default service config");
+                    "Fallback to error due to invalid first service config without default config");
+                onError(configOrError.getError());
+                return;
+              } else {
+                effectiveServiceConfig = lastServiceConfig;
               }
+            } else {
+              effectiveServiceConfig = EMPTY_SERVICE_CONFIG;
             }
-
             // FIXME(notcarl): reference equality is not right (although not harmful) right now.
             //                 Name resolver should return the same config if txt record is the same
             if (effectiveServiceConfig != lastServiceConfig) {
-              channelLogger.log(ChannelLogLevel.INFO,
-                  "Service config changed{0}", effectiveServiceConfig == null ? " to null" : "");
+              channelLogger.log(
+                  ChannelLogLevel.INFO,
+                  "Service config changed{0}",
+                  effectiveServiceConfig == EMPTY_SERVICE_CONFIG ? " to empty" : "");
               lastServiceConfig = effectiveServiceConfig;
             }
 
-            try {
-              handleServiceConfigUpdate();
-            } catch (RuntimeException re) {
-              logger.log(
-                  Level.WARNING,
-                  "[" + getLogId() + "] Unexpected exception from parsing service config",
-                  re);
+            if (!(servers.isEmpty() && lastResolutionStateCopy == ResolutionState.SUCCESS)) {
+              try {
+                handleServiceConfigUpdate();
+              } catch (RuntimeException re) {
+                logger.log(
+                    Level.WARNING,
+                    "[" + getLogId() + "] Unexpected exception from parsing service config",
+                    re);
+              }
             }
           }
 
           // Call LB only if it's not shutdown.  If LB is shutdown, lbHelper won't match.
           if (NameResolverListener.this.helper == ManagedChannelImpl.this.lbHelper) {
             Attributes effectiveAttrs = attrs;
-            if (effectiveServiceConfig != serviceConfig) {
+            if (effectiveServiceConfig != validServiceConfig
+                && !effectiveServiceConfig.rawServiceConfig.isEmpty()) {
               effectiveAttrs = attrs.toBuilder()
-                  .set(GrpcAttributes.NAME_RESOLVER_SERVICE_CONFIG, effectiveServiceConfig)
+                  .set(
+                      GrpcAttributes.NAME_RESOLVER_SERVICE_CONFIG,
+                      effectiveServiceConfig.rawServiceConfig)
                   .build();
             }
+
+            Object loadBalancingPolicyConfig = effectiveServiceConfig.parsed.getConfig();
```

Comment:
You should extract `PolicySelection` out of this `ManagedChannelServiceConfig` and pass it to `.setLoadBalancingPolicyConfig()`, because `lb` returned a `PolicySelection` from `parseLoadBalancingPolicy()`.